builtin/gc.c

```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
 	return 0;
 }
 
+static void link_or_copy_or_die(const char *src, const char *dst)
+{
+	if (!link(src, dst))
+		return;
+
+	warning_errno(_("failed to link '%s' to '%s'"), src, dst);
+
+	if (!copy_file(dst, src, 0644))
+		return;
+
+	die_errno(_("failed to copy '%s' to '%s'"), src, dst);
+}
+
+static void migrate_pack(const char *srcdir, const char *dstdir,
+			 const char *pack_filename)
+{
+	struct stat st;
+	char *basename;
+	struct strbuf src = STRBUF_INIT, dst = STRBUF_INIT;
+	struct {
+		const char *ext;
+		unsigned move:1;
+	} files[] = {
+		{".pack", 0},
+		{".keep", 0},
+		{".rev", 0},
+		{".idx", 1}, /* The index file must be atomically moved last. */
+	};
+
+	trace2_region_enter("maintenance", "migrate_pack", the_repository);
+
+	basename = xstrndup(pack_filename, strlen(pack_filename) - 5 /*.pack*/);
+
+	/* A pack without an index file is not yet ready to be migrated. */
+	strbuf_addf(&src, "%s/%s%s", srcdir, basename, ".idx");
+	if (stat(src.buf, &st))
+		goto cleanup;
+
+	/* Move or copy files from the source directory to the destination. */
+	for (size_t i = 0; i < ARRAY_SIZE(files); i++) {
+		strbuf_reset(&src);
+		strbuf_addf(&src, "%s/%s%s", srcdir, basename, files[i].ext);
+
+		if (stat(src.buf, &st))
+			continue;
+
+		strbuf_reset(&dst);
+		strbuf_addf(&dst, "%s/%s%s", dstdir, basename, files[i].ext);
+
+		if (files[i].move) {
+			if (rename(src.buf, dst.buf))
+				die_errno(_("failed to move '%s' to '%s'"),
+					  src.buf, dst.buf);
+		} else {
+			link_or_copy_or_die(src.buf, dst.buf);
+		}
+	}
+
+	/*
+	 * Now the pack and all associated files exist at the destination we can
+	 * now clean up the files in the source directory.
+	 */
+	for (size_t i = 0; i < ARRAY_SIZE(files); i++) {
+		/* Files that were moved rather than copied have no clean up. */
+		if (files[i].move)
+			continue;
+
+		strbuf_reset(&src);
+		strbuf_addf(&src, "%s/%s%s", srcdir, basename, files[i].ext);
+		if (unlink(src.buf))
+			warning_errno(_("failed to delete '%s'"), src.buf);
+	}
+
+cleanup:
+	free(basename);
+	strbuf_release(&src);
+	strbuf_release(&dst);
+
+	trace2_region_leave("maintenance", "migrate_pack", the_repository);
+}
+
+static void move_pack_to_vfs_cache(const char *full_path, size_t full_path_len,
+				   const char *file_name, UNUSED void *data)
```

Comment:
It might make sense to merge `migrate_pack()` directly into this function, after the `.pack` suffix check.